README.md
# React-draggable-tab [![Build Status](https://travis-ci.org/georgeOsdDev/react-draggable-tab.svg?branch=develop)](https://travis-ci.org/georgeOsdDev/react-draggable-tab) [![npm version](https://badge.fury.io/js/react-draggable-tab.svg)](http://badge.fury.io/js/react-draggable-tab)


[![Gyazo](http://i.gyazo.com/42d408d288292f62fbb8d650897acbc4.gif)](http://gyazo.com/42d408d288292f62fbb8d650897acbc4)

Atom like draggable tab react component.

## Demo

[View Demo](http://georgeosddev.github.io/react-draggable-tab/example/)

## Installation

```bash
npm install --save react-draggable-tab
```
React v0.14 is supported from react-draggable-tab v0.4.0.
For React v0.13.x, please use react-draggable-tab v0.3.3.

## API

### `Tab`

`Tab` is just a case class to check props.
`props.children` will rendered into content area.

#### Props

  * `key`: *unique* key in `TabList`.
    `PropTypes.string.isRequired`

  * `beforeTitle`: element to show in tab. eg icon.
    `PropTypes.element`

  * `title`: string or element to show in tab.
    `PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired`

  * `afterTitle`: element to show in tab. eg: notification badge
    `PropTypes.element`

  * `uncloseable`: If `true`, closeButton will not be appeared in tab.
    `PropTypes.bool` (Default `false`)

###### Style (for each tab)

  * `tabClassNames`: classNames which will be **added** to rendered elements.
   * `tab`: base `li` element of tab (defult: `rdTab`)
   * `tabBefore`: before element of `li` which emulate `:Before` selector (defult: `rdTabBefore`)
   * `tabAfter`: after element of `li` which emulate `:After` selector (defult: `rdTabAfter`)
   * `tabTitle`: `span` element of tab title (defult: `rdTabTitle`)
   * `tabBeforeTitle`: `span` element of tab before title (defult: `tabBeforeTitle`)
   * `tabAfterTitle`: `span` element of tab after title (defult: `tabAfterTitle`)
   * `tabCloseIcon`: base `span` element of close icon (defult: `rdCloseIcon`)
   * `tabActive`: selected tab's `li`, before, after (defult: `rdTabActive`)
   * `tabHover`: selected tab's `li`, before, after (defult: `rdTabHover`)

  * `tabStyles`: Inline styles which will be **overwritten** default and common-tabs inline styles.
   * `tab`: base `li` element of tab
   * `tabBefore`: before element of `li` which emulate `:Before` selector.
   * `tabAfter`: after element of `li` which emulate `:After` selector.
   * `tabTitle`: `span` element of tab title
   * `tabActive`: selected tab's `li`
   * `tabBeforeActive`: selected tab's `li` before
   * `tabAfterActive`: selected tab's `li` after
   * `tabTitleActive`: selected tab's title
   * `tabOnHover`: hovered tab's `li`
   * `tabBeforeOnHover`: hovered tab's `li` before
   * `tabAfterOnHover`: hovered tab's `li` after
   * `tabTitleOnHover`: hovered tab's title
   * `tabCloseIcon`: base `span` element of close icon
   * `tabCloseIconOnHover`: base `span` element of close icon when hover

  * `containerStyle`: style object which will be apply to container of rendered tab.

  * `hiddenContainerStyle`: style object which will be apply to container of hidden tab.


##### Events

  All other props like `onXX` handler set to `Tab` will be passed to rendered element except `onClick`
  You can use any `onXX` for [Supported events](https://facebook.github.io/react/docs/events.html#supported-events) for tab element.

### `Tabs`

`Tabs` is container for tab. it will render tabBar and content of selected element.

#### Props

  * `tabs`: Array of `Tab` elements.
    `PropTypes.arrayOf(PropTypes.element)`

  * `selectedTab`: key for selectedTab.
    `PropTypes.string` default to first tab.

  * `disableDrag`: disables the ability to drag the tabs.
    `React.PropTypes.bool` default is false.

  * `tabAddButton`: element for add button.
    `PropTypes.element`

  * `keepSelectedTab`: Prevent tab select on drag/move behind tab.
    `PropTypes.bool` default `false`.

  * `unclosable`: Disable tab to close.
    `PropTypes.bool` default `false`.

  * `shouldTabClose(e, key)`: will be called before tab close event, return false if you want to stop tab close process, default `true`;

###### Shortcut key binding
  * `shortCutKeys`: Short cut key bindings as [Mousetrap](https://craig.is/killing/mice) style.
   * `close`: key binding to close current tab (`onTabClose` will be called)
   * `create`: key binding to create tab (`onTabAddButtonClick` will be called)
   * `moveRight`: key binding to move right (`onTabSelect` will be called)
   * `moveLeft`: key binding to move left (`onTabSelect` will be called)

###### Style (All tabs will be apply these styles)

  * `tabsClassNames`: classNames which will be **added** to rendered elements.
   * `tabWrapper`: root wrapper `div` element (defult: `rdTabWrapper`)
   * `tabBar`: base `ul` element of tab bar (defult: `rdTabBar`)
   * `tabBarAfter`: after `span` element of tab bar which emulate `:After` selector (defult: `rdTabBarAfter`)
   * `tab`: base `li` element of tab (defult: `rdTab`)
   * `tabBefore`: before element of `li` which emulate `:Before` selector (defult: `rdTabBefore`)
   * `tabAfter`: after element of `li` which emulate `:After` selector (defult: `rdTabAfter`)
   * `tabTitle`: `span` element of tab title (defult: `rdTabTitle`)
   * `tabBeforeTitle`: `span` element of tab before title (defult: `rdTabBeforeTitle`)
   * `tabBeforeTitle`: `span` element of tab after title (defult: `rdTabAfterTitle`)
   * `tabCloseIcon`: base `span` element of close icon (defult: `rdCloseIcon`)
   * `tabActive`: selected tab's `li`, before, after (defult: `rdTabActive`)

  * `tabsStyles`: Inline styles which will be **overwritten** default inline styles.
   * `tabWrapper`: root wrapper `div` element
   * `tabBar`: base `ul` element of tab bar
   * `tabBarAfter`: after `span` element of tab bar which emulate `:After` selector
   * `tab`: base `li` element of tab
   * `tabBefore`: before element of `li` which emulate `:Before` selector.
   * `tabAfter`: after element of `li` which emulate `:After` selector.
   * `tabTitle`: `span` element of tab title
   * `tabActive`: selected tab's `li`
   * `tabBeforeActive`: selected tab's `li` before
   * `tabAfterActive`: selected tab's `li` after
   * `tabTitleActive`: selected tab's title
   * `tabCloseIcon`: base `span` element of close icon
   * `tabCloseIconOnHover`: base `span` element of close icon when hover

##### Events

  * `onTabSelect(e, key, currentTabs)`: Called when tab of key was selected.
    `currentTabs` is array of tabs elements sorted with current order.

  * `onTabClose(e, key, currentTabs)`: Called when tab of key was closed.
    `currentTabs` is array of tabs elements sorted with current order.

  * `onTabPositionChange(e, key, currentTabs)`: Called when tab of key was moved.
    `currentTabs` is array of tabs elements sorted with current order.

  * `onTabAddButtonClick(e, currentTabs)`: Called when `tab add button` was clicked.
    `currentTabs` is array of tabs elements sorted with current order.
    Basically you will concat `currentTabs` with new empty tab.

    ```javascript
    let newTabs = currentTabs.concat([newTab]);
    ```

  ~~* `onTabDoubleClick(e, key)`: Called when `title` was double clicked.~~ Removed from v0.5.0

  ~~* `onTabMouseEnter(e, key)`: Called when mouse enter to `tab`.~~ Removed from v0.5.0

  ~~* `onTabMouseLeave(e, key)`: Called when mouse leave from `tab`.~~ Removed from v0.5.0

## Usage example

```javascript
class App extends React.Component {
  constructor(props) {
    super(props);
    let icon = (<image src='icon.png' style={{height:'13px'}}/>);
    let fonticon = (<icon className='icon-html5'/>);
    let badge = (<DynamicTabBadge />);

    this.state = {
      tabs:[
        (<Tab key={'tab0'} title={'unclosable tab'} unclosable={true} >
          <div>
            <h1>This tab cannot close</h1>
          </div>
        </Tab>),
        (<Tab key={'tab1'} title={'1stTab'} beforeTitle={icon} >
          <div>
            <h1>This is tab1</h1>
          </div>
        </Tab>),
        (<Tab key={'tab2'} title={'2ndTab Too long Toooooooooooooooooo long'} beforeTitle={fonticon} >
          <div>
            <pre>Lorem ipsum dolor sit amet, consectetur adipisicing elit,
            </pre>
          </div>
        </Tab>),
        (<Tab key={'tab3'} title={'Dynamic tab'} afterTitle={badge}>
          <DynamicTabContent/>
        </Tab>)
      ],
      badgeCount: 0
    };
  }

  handleTabSelect(e, key, currentTabs) {
    console.log('handleTabSelect key:', key);
    this.setState({selectedTab: key, tabs: currentTabs});
  }

  handleTabClose(e, key, currentTabs) {
    console.log('tabClosed key:', key);
    this.setState({tabs: currentTabs});
  }

  handleTabPositionChange(e, key, currentTabs) {
    console.log('tabPositionChanged key:', key);
    this.setState({tabs: currentTabs});
  }

  handleTabAddButtonClick(e, currentTabs) {
    // key must be unique
    const key = 'newTab_' + Date.now();
    let newTab = (<Tab key={key} title='untitled'>
                    <div>
                      <h1>New Empty Tab</h1>
                    </div>
                  </Tab>);
    let newTabs = currentTabs.concat([newTab]);

    this.setState({
      tabs: newTabs,
      selectedTab: key
    });
  }

  render() {

    return (
      <Tabs
        tabsClassNames={tabsClassNames}
        tabsStyles={tabsStyles}
        selectedTab={this.state.selectedTab ? this.state.selectedTab : "tab2"}
        onTabSelect={this.handleTabSelect.bind(this)}
        onTabClose={this.handleTabClose.bind(this)}
        onTabAddButtonClick={this.handleTabAddButtonClick.bind(this)}
        onTabPositionChange={this.handleTabPositionChange.bind(this)}
        tabs={this.state.tabs}
        shortCutKeys={
          {
            'close': ['alt+command+w', 'alt+ctrl+w'],
            'create': ['alt+command+t', 'alt+ctrl+t'],
            'moveRight': ['alt+command+tab', 'alt+ctrl+tab'],
            'moveLeft': ['shift+alt+command+tab', 'shift+alt+ctrl+tab']
          }
        }
      />
    )
  }
};
```

See also [example](https://github.com/georgeOsdDev/react-draggable-tab/tree/develop/example)


```bash
npm install
npm run start:example
```

## Tests

```bash
npm test
```

## Contributors

 See [list](https://github.com/georgeOsdDev/react-draggable-tab/graphs/contributors)

## Known Issue

* Dynamic tab content.

`Tabs` do not care any change in `Tab` content.
content needs update by your application side.
See `3rdTab` in example.


* `flex` style should be define in CSS for safari.
See https://github.com/facebook/react/issues/2020

In application, class `rdTabBar` or your custom class of `TabBar` needs `display: -webkit-flex` in CSS like below.
```css
.myTabBar {
  display: -webkit-flex;
}



CHANGELOG.md
## Change Log

### Ver 0.10.1

  * #101 [Received `false` for a non-boolean attribute `uncloseable`](https://github.com/georgeOsdDev/react-draggable-tab/issues/101)
### Ver 0.10.0

  * #99 [Fix this.unbindShortcuts is not a function](https://github.com/georgeOsdDev/react-draggable-tab/issues/99)
  * Fix falling unit tests

### Ver 0.9.0

  This version contains a breaking change. The name of the Tab 'disableClose' property,
  which shadowed a standard property name and resulted in a React warning, has been changed
  to 'unclosable'.

  **TESTS ARE BROKEN**. `react-addons-test-utils` has a hard dependency on React 15. For
  React 16, Facebook has moved the utilities into `react-dom/test-utils` and
  `react-test-renderer/shallow`, and they are quite a bit different.

  * #73 [Invariant Violation: addComponentAsRefTo(...): ...](https://github.com/georgeOsdDev/react-draggable-tab/issues/73)
  may be better. The React version has been updated, but react-draggable-tab is still using string refs, which is deprecated.
  * #90 [React.PropTypes should be replaced by just PropTypes](https://github.com/georgeOsdDev/react-draggable-tab/issues/90)
  * #92 [Update to React v16](https://github.com/georgeOsdDev/react-draggable-tab/issues/92)
  Updated all libraries. Uses React 16.2.0.
  * #93 [Resolve all eslint "airbnb" errors](https://github.com/georgeOsdDev/react-draggable-tab/issues/93)
  (except in tests)

### Ver 0.8.1

  * #80 [Fix tabAddButton position](https://github.com/georgeOsdDev/react-draggable-tab/issues/80)
  Thanks @warent
  * #81 [Added prop to disable the dragging of tabs]()https://github.com/georgeOsdDev/react-draggable-tab/issues/81
  Thanks @erangil2

### Ver 0.8.0

  * #77 [Add shouldTabClose](https://github.com/georgeOsdDev/react-draggable-tab/issues/77)

### Ver 0.7.0

  * #67 [Fix TypeError when using SVGs as title](https://github.com/georgeOsdDev/react-draggable-tab/issues/67)
  Thanks @HaNdTriX
  * #68 [Improve tab rendering perfomance](https://github.com/georgeOsdDev/react-draggable-tab/issues/68)

### Ver 0.6.1

  * #66 [Removes console warning in React15](https://github.com/georgeOsdDev/react-draggable-tab/issues/66)
    Thanks @uriklar

### Ver 0.6.0

  * #63 [React v15 support](https://github.com/georgeOsdDev/react-draggable-tab/issues/63)

### Ver 0.5.1
  * #59 [Allow overriding TabStyles.wrapper using props](https://github.com/georgeOsdDev/react-draggable-tab/issues/59)

### Ver 0.5.0
  * #53 [Context menu on right click](https://github.com/georgeOsdDev/react-draggable-tab/issues/53)
    `onTabDoubleClick`, `onTabMouseEnter` and `onTabMouseEnter` are removed from `Tabs`.
    All `onXXX` handler except `onClick` are now used on `Tab`.
  * Update dependencies libraries version [cf65dbc](https://github.com/georgeOsdDev/react-draggable-tab/commit/cf65dbc8f756561536f53f5e3960bf86afebdc73)

### Ver 0.4.3

  * #48 [When tab title is truncated onhover does not show whole title](https://github.com/georgeOsdDev/react-draggable-tab/issues/48)
  * #50 [Unable to add styles or css classes to the tab content div](https://github.com/georgeOsdDev/react-draggable-tab/issues/50)

### Ver 0.4.2

  * #45 [Unexpected diff between src and lib](https://github.com/georgeOsdDev/react-draggable-tab/issues/45)

### Ver 0.4.1

  * #33 [Add hover effect on mouseOver like chrome](https://github.com/georgeOsdDev/react-draggable-tab/issues/33)
  * #38 [Publish transpiled code](https://github.com/georgeOsdDev/react-draggable-tab/issues/38)
  * #40 [Tab start moving from strange position](https://github.com/georgeOsdDev/react-draggable-tab/issues/33)

### Ver 0.4.0

  This version updated react from ^0.13.3 to 0.14.X, react-draggable from ^0.8.0 to 1.1.X, and also other dependencies.
  Several changes have been made for adopting react v0.14.x.

  * #35 [Update dependencies](https://github.com/georgeOsdDev/react-draggable-tab/issues/35)
  * #32 [Add `keepSelectedTab` option](https://github.com/georgeOsdDev/react-draggable-tab/issues/32)

### Ver 0.3.3

  * #30 [Switch tab position while dragging](https://github.com/georgeOsdDev/react-draggable-tab/issues/30)

### Ver 0.3.2

  * #27 [Support keyboard shortcut](https://github.com/georgeOsdDev/react-draggable-tab/issues/27)

### Ver 0.3.1

  * #26 [Enable to use element for tab title](https://github.com/georgeOsdDev/react-draggable-tab/issues/26)

### Ver 0.3.0

  This release contains breaking change.

  * #20 [Remove 'ed' from event handler name](https://github.com/georgeOsdDev/react-draggable-tab/issues/20) (Breaking change)
  * #21 [Unexpected behavior with NODE_ENV=production](https://github.com/georgeOsdDev/react-draggable-tab/issues/21)

### Ver 0.2.3

  * #16 [Add doubleClick handler on tab title](https://github.com/georgeOsdDev/react-draggable-tab/issues/16)

### Ver 0.2.2

  * #13 [Improve tabBarAfter style](https://github.com/georgeOsdDev/react-draggable-tab/issues/13)

### Ver 0.2.1

  * #5 [Enable to add badge / favicon like chrome](https://github.com/georgeOsdDev/react-draggable-tab/issues/5)
  * #6 [Add style on parent div](https://github.com/georgeOsdDev/react-draggable-tab/issues/6)
  * #8 [Show closed tab when it supplied again with same key](https://github.com/georgeOsdDev/react-draggable-tab/issues/8)
  * #9 [Keep all tab content inside Tabs](https://github.com/georgeOsdDev/react-draggable-tab/issues/9)

### Ver 0.2.0

  * #1 [Improve tab title color](https://github.com/georgeOsdDev/react-draggable-tab/issues/1)
  * #2 [onTabSelected, onTabClosed, onTabPositionChanged should be called in setState callback](https://github.com/georgeOsdDev/react-draggable-tab/issues/2)
  * #3 [Enable to set style and class on each Tab](https://github.com/georgeOsdDev/react-draggable-tab/issues/3)

### Ver 0.1.0 Initial release


example/app.js
'use strict';

import _ from 'lodash';
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import {Dialog, FlatButton, Menu, MenuItem, TextField} from 'material-ui';
import {Tabs, Tab} from '../lib/index.js';

import DynamicTabContent from './DynamicTabContent';
import DynamicTabBadge from './DynamicTabBadge';

//allow react dev tools work
window.React = React;

const tabsClassNames = {
  tabWrapper: 'myWrapper',
  tabBar: 'myTabBar',
  tabBarAfter: 'myTabBarAfter',
  tab:      'myTab',
  tabTitle: 'myTabTitle',
  tabCloseIcon: 'tabCloseIcon',
  tabBefore: 'myTabBefore',
  tabAfter: 'myTabAfter'
};

const tabsStyles = {
  tabWrapper: {marginTop: '10px'},
  tabBar: {},
  tab:{},
  tabTitle: {},
  tabCloseIcon: {},
  tabBefore: {},
  tabAfter: {}
};

class App extends React.Component {
  constructor(props) {
    super(props);

    let icon = (<img src='icon.png' style={{height:'13px'}}/>);
    let fonticon = (<span className='icon-html5'/>);
    let badge = (<DynamicTabBadge />);

    this.state = {
      tabs:[
        (<Tab key={'tab0'} title={'unclosable tab'} unclosable={true} {...this.makeListeners('tab0')}>
          <div>
            <h1>This tab cannot close</h1>
          </div>
        </Tab>),
        (<Tab key={'tab1'} title={'1stTab'} beforeTitle={icon} {...this.makeListeners('tab1')}>
          <div>
            <h1>This is tab1</h1>
          </div>
        </Tab>),
        (<Tab key={'tab2'} title={'2ndTab Too long Toooooooooooooooooo long'} beforeTitle={fonticon} {...this.makeListeners('tab2')}>
          <div>
            <pre>Lorem ipsum dolor sit amet, consectetur adipisicing elit,
            </pre>
          </div>
        </Tab>),
        (<Tab key={'tab3'} title={'Dynamic tab'} afterTitle={badge} {...this.makeListeners('tab3')}>
          <DynamicTabContent/>
        </Tab>),
        (<Tab key={'tab4'} title={'Custom container'} containerStyle={{backgroundColor: 'gray', width: '50%'}} {...this.makeListeners('tab4')}>
          <div>
            <h1>This is tab4 with custom container style</h1>
          </div>
        </Tab>),
        (<Tab key={'tab5'} title={'Big content1 with left:-9999999px'} hiddenContainerStyle={{left: '-9999999px', top: '-9999999px'}} {...this.makeListeners('tab5')}>
          <div>
            <h1>Super big content</h1>
              {Array(10000).fill(0).map((_, i) => <p key={i}>Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>)}
          </div>
        </Tab>),
        (<Tab key={'tab6'} title={'after big content'} {...this.makeListeners('tab6')}>
          <div>
            <pre>Lorem ipsum dolor sit amet, consectetur adipisicing elit,
            </pre>
          </div>
        </Tab>),
      ],
      badgeCount: 0,
      menuPosition: {},
      showMenu: false,
      dialogOpen: false
    };
  }

  getChildContext() {
    return { muiTheme: getMuiTheme()};
  }

  // getChildContext() {
  //   return {
  //     muiTheme: ThemeManager.getMuiTheme()
  //   };
  // }

  makeListeners(key){
    return {
      onClick: (e) => { console.log('onClick', key, e);}, // never called
      onContextMenu: (e) => { console.log('onContextMenu', key, e); this.handleTabContextMenu(key, e)},
      onDoubleClick: (e) => { console.log('onDoubleClick', key, e); this.handleTabDoubleClick(key, e)},
    }
  }

  handleTabSelect(e, key, currentTabs) {
    console.log('handleTabSelect key:', key);
    this.setState({selectedTab: key, tabs: currentTabs});
  }

  handleTabClose(e, key, currentTabs) {
    console.log('tabClosed key:', key);
    this.setState({tabs: currentTabs});
  }

  handleTabPositionChange(e, key, currentTabs) {
    console.log('tabPositionChanged key:', key);
    this.setState({tabs: currentTabs});
  }

  handleTabAddButtonClick(e, currentTabs) {
    // key must be unique
    const key = 'newTab_' + Date.now();
    let newTab = (<Tab key={key} title='untitled' {...this.makeListeners(key)}>
                    <div>
                      <h1>New Empty Tab</h1>
                    </div>
                  </Tab>);
    let newTabs = currentTabs.concat([newTab]);

    this.setState({
      tabs: newTabs,
      selectedTab: key
    });
  }

  handleTabDoubleClick(key) {
    this.setState({
      editTabKey: key,
      dialogOpen: true
    });
  }

  handleTabContextMenu(key, e) {
    e.preventDefault();
    this.setState({
      showMenu: true,
      contextTarget: key,
      menuPosition: {
        top:`${e.pageY}px`,
        left:`${e.pageX}px`
      }
    });
  }

  cancelContextMenu(){
    this.setState({
      showMenu: false,
      contextTarget: null
    });
  }

  renameFromContextMenu(){
    this.setState({
      showMenu: false,
      contextTarget: null,
      editTabKey: this.state.contextTarget,
      dialogOpen: true
    });
  }

  closeFromContextMenu(){
    let newTabs = _.filter(this.state.tabs, (t) => {return t.key !== this.state.contextTarget;});
    this.setState({
      showMenu: false,
      contextTarget: null,
      selectedTab: 'tab0',
      tabs: newTabs
    });
  }

  _onDialogSubmit() {
    const replaceFunc = _.bind((tab) => {
        if (tab.key === this.state.editTabKey) {
          return React.cloneElement(tab, {title: this.refs.input.getValue()});
        }
        return tab;
      }, this),
      newTabs = _.map(this.state.tabs, replaceFunc);
    this.setState({
      tabs: newTabs,
      dialogOpen: false
    });
  }

  _onDialogCancel() {
    this.setState({
      dialogOpen: false
    })
  }

  _onDialogShow() {
    let tab = _.find(this.state.tabs, (t) => {
      return t.key === this.state.editTabKey;
    });
    this.refs.input.setValue(tab.props.title);
  }

  shouldTabClose(e, key){
    console.log('should tab close', e, key);
    return window.confirm('close?');
  }

  render() {

    let standardActions = [
      <FlatButton
        label="Cancel"
        secondary={true}
        onClick={this._onDialogCancel.bind(this)}
      />,
      <FlatButton
        label="Submit"
        primary={true}
        keyboardFocused={true}
        onClick={this._onDialogSubmit.bind(this)}
      />
    ];

    let menuStyle = {
      display: this.state.showMenu ? 'block': 'none',
      position: 'absolute',
      top: this.state.menuPosition.top,
      left: this.state.menuPosition.left,
      backgroundColor: '#F0F0F0'
    };

    return (
      <div>
        <Tabs
          tabsClassNames={tabsClassNames}
          tabsStyles={tabsStyles}
          selectedTab={this.state.selectedTab ? this.state.selectedTab : 'tab2'}
          onTabSelect={this.handleTabSelect.bind(this)}
          onTabClose={this.handleTabClose.bind(this)}
          onTabAddButtonClick={this.handleTabAddButtonClick.bind(this)}
          onTabPositionChange={this.handleTabPositionChange.bind(this)}
          shouldTabClose={this.shouldTabClose.bind(this)}
          tabs={this.state.tabs}
          shortCutKeys={
            {
              'close': ['alt+command+w', 'alt+ctrl+w'],
              'create': ['alt+command+t', 'alt+ctrl+t'],
              'moveRight': ['alt+command+tab', 'alt+ctrl+tab'],
              'moveLeft': ['shift+alt+command+tab', 'shift+alt+ctrl+tab']
            }
          }
          keepSelectedTab={true}
        />
        <div style={menuStyle}>
          <Menu>
            {this.state.contextTarget === 'tab0' ? '' : <MenuItem primaryText="Close" onClick={this.closeFromContextMenu.bind(this)}/>}
            <MenuItem primaryText="Rename" onClick={this.renameFromContextMenu.bind(this)}/>
            <MenuItem primaryText="Cancel" onClick={this.cancelContextMenu.bind(this)}/>
          </Menu>
        </div>
        <Dialog
          title="Change tab name"
          ref="dialog"
          actions={standardActions}
          modal={true}
          open={this.state.dialogOpen}
          onShow={this._onDialogShow.bind(this)}>
          <TextField
            ref='input' id="rename-input" style={{width: '90%'}}/>
        </Dialog>
        <p style={{position: 'fixed', 'bottom': '10px'}}>
          Source code can be found at <a href='https://github.com/georgeOsdDev/react-draggable-tab/tree/master/example'>GitHub</a>
        </p>
      </div>
    );
  }
}

App.childContextTypes = {
  muiTheme: PropTypes.object
};

ReactDOM.render(<App/>, document.getElementById('tabs'));
